I'm able to reproduce locally. The thread local variable is working how I expect, but the test is making an assumption that the DataFrame doesn't have a `query_job` populated. This assumption is violated when the DataFrame is used across multiple tests.

I'll update the test.

```
________________________________________________________________________________ test_query_job_dry_run ________________________________________________________________________________
[gw2] linux -- Python 3.11.1 /usr/local/google/home/swast/envs/bigframes/bin/python3.11

penguins_df_default_index =                                       species     island  culmen_length_mm  \
0           Gentoo penguin (Pygoscelis p...  <NA>         <NA>    <NA>  
24             20.7              191.0       3900.0  FEMALE  
...

[344 rows x 7 columns]

    def test_query_job_dry_run(penguins_df_default_index: bf.dataframe.DataFrame):
        expected = "Computation deferred. Computation will process"
    
        with bf.option_context("display.repr_mode", "deferred"):
            df_result = repr(penguins_df_default_index)
>           assert expected in df_result
E           AssertionError: assert 'Computation deferred. Computation will process' in 'Query Job Info\nJob url: https://console.cloud.google.com/bigquery?project=swast-scratch&j=bq:US:9a761424-b164-4c31-a...240503_session444e7b_fbabcb4962f841a8bb92ec2f27d17bb2\nSlot Time: a moment\nBytes Processed: 28.9 kB\nCache hit: False'

tests/system/small/test_progress_bar.py:142: AssertionError
--------------------------------------------------------------------------------- Captured stdout call ---------------------------------------------------------------------------------
DisplayOptions(max_columns=20, max_rows=25, progress_bar='auto', repr_mode='deferred', max_info_columns=100, max_info_rows=200000, memory_usage=True)
deferred
```

--- a/bigframes/_config/__init__.py
+++ b/bigframes/_config/__init__.py
@@ -44,7 +44,6 @@ class Options:
         # BigQuery options are special because they can only be set once per
         # session, so we need an indicator as to whether we are using the
         # thread-local session or the global session.
-        self._local.is_bigquery_thread_local = False
         self._bigquery_options = bigquery_options.BigQueryOptions()
 
     def _init_bigquery_thread_local(self):
@@ -53,20 +52,18 @@ class Options:
         # Already thread-local, so don't reset any options that have been set
         # already. No locks needed since this only modifies thread-local
         # variables.
-        if self._local.is_bigquery_thread_local:
+        if self._local.bigquery_options is not None:
             return
 
-        self._local.is_bigquery_thread_local = True
         self._local.bigquery_options = copy.deepcopy(self._bigquery_options)
         self._local.bigquery_options._session_started = False
 
     @property
     def bigquery(self) -> bigquery_options.BigQueryOptions:
         """Options to use with the BigQuery engine."""
-        if self._local.is_bigquery_thread_local:
+        if self._local.bigquery_options is not None:
             # The only way we can get here is if someone called
-            # _init_bigquery_thread_local, which initializes
-            # self._local.bigquery_options.
+            # _init_bigquery_thread_local.
             return self._local.bigquery_options
 
         return self._bigquery_options
@@ -106,9 +103,10 @@ class Options:
     def is_bigquery_thread_local(self) -> bool:
         """Indicator that we're using a thread-local session.
 
-        This is set to True by using `option_context` with a "bigquery" option.
+        A thread-local session can be started by using
+        `with bigframes.option_context("bigquery.some_option", "some-value"):`.
         """
-        return self._local.is_bigquery_thread_local
+        return self._local.bigquery_options is not None
 
 
 options = Options()

--- a/bigframes/core/global_session.py
+++ b/bigframes/core/global_session.py
@@ -63,9 +63,10 @@ def close_session() -> None:
             _global_session_state.thread_local_session = None
 
         # Currently using thread-local options, so no global lock needed.
-        # Don't reset is_bigquery_thread_local, as that's the responsibility
+        # Don't reset options.bigquery, as that's the responsibility
         # of the context manager that started it in the first place. The user
-        # might have explicitly closed the session in the context manager.
+        # might have explicitly closed the session in the context manager and
+        # the thread-locality property needs to be retained.
         bigframes._config.options.bigquery._session_started = False
 
         # Don't close the non-thread-local session.

--- a/bigframes/core/indexes/base.py
+++ b/bigframes/core/indexes/base.py
@@ -239,7 +239,7 @@ class Index(vendored_pandas_index.Index):
         opts = bigframes.options.display
         max_results = opts.max_rows
         if opts.repr_mode == "deferred":
-            return formatter.repr_query_job(self.query_job)
+            return formatter.repr_query_job(self._block._compute_dry_run())
 
         pandas_df, _, query_job = self._block.retrieve_repr_request_results(max_results)
         self._query_job = query_job

--- a/bigframes/dataframe.py
+++ b/bigframes/dataframe.py
@@ -595,7 +595,7 @@ class DataFrame(vendored_pandas_frame.DataFrame):
         opts = bigframes.options.display
         max_results = opts.max_rows
         if opts.repr_mode == "deferred":
-            return formatter.repr_query_job(self.query_job)
+            return formatter.repr_query_job(self._compute_dry_run())
 
         self._cached()
         # TODO(swast): pass max_columns and get the true column count back. Maybe
@@ -634,7 +634,7 @@ class DataFrame(vendored_pandas_frame.DataFrame):
         opts = bigframes.options.display
         max_results = opts.max_rows
         if opts.repr_mode == "deferred":
-            return formatter.repr_query_job_html(self.query_job)
+            return formatter.repr_query_job(self._compute_dry_run())
 
         self._cached()
         # TODO(swast): pass max_columns and get the true column count back. Maybe

--- a/bigframes/series.py
+++ b/bigframes/series.py
@@ -282,7 +282,7 @@ class Series(bigframes.operations.base.SeriesMethods, vendored_pandas_series.Ser
         opts = bigframes.options.display
         max_results = opts.max_rows
         if opts.repr_mode == "deferred":
-            return formatter.repr_query_job(self.query_job)
+            return formatter.repr_query_job(self._compute_dry_run())
 
         self._cached()
         pandas_df, _, query_job = self._block.retrieve_repr_request_results(max_results)

--- a/tests/system/small/test_progress_bar.py
+++ b/tests/system/small/test_progress_bar.py
@@ -23,6 +23,7 @@ import bigframes.formatting_helpers as formatting_helpers
 from bigframes.session import MAX_INLINE_DF_BYTES
 
 job_load_message_regex = r"\w+ job [\w-]+ is \w+\."
+EXPECTED_DRY_RUN_MESSAGE = "Computation deferred. Computation will process"
 
 
 def test_progress_bar_dataframe(
@@ -134,13 +135,19 @@ def test_query_job_repr(penguins_df_default_index: bf.dataframe.DataFrame):
         assert string in query_job_repr
 
 
-def test_query_job_dry_run(penguins_df_default_index: bf.dataframe.DataFrame):
-    expected = "Computation deferred. Computation will process"
-
+def test_query_job_dry_run_dataframe(penguins_df_default_index: bf.dataframe.DataFrame):
     with bf.option_context("display.repr_mode", "deferred"):
         df_result = repr(penguins_df_default_index)
-        assert expected in df_result
+        assert EXPECTED_DRY_RUN_MESSAGE in df_result
 
+
+def test_query_job_dry_run_index(penguins_df_default_index: bf.dataframe.DataFrame):
+    with bf.option_context("display.repr_mode", "deferred"):
+        index_result = repr(penguins_df_default_index.index)
+        assert EXPECTED_DRY_RUN_MESSAGE in index_result
+
+
+def test_query_job_dry_run_series(penguins_df_default_index: bf.dataframe.DataFrame):
     with bf.option_context("display.repr_mode", "deferred"):
         series_result = repr(penguins_df_default_index["body_mass_g"])
-        assert expected in series_result
+        assert EXPECTED_DRY_RUN_MESSAGE in series_result

